### PR TITLE
editoast: create `editoast_core` crate and remove `core` module

### PR DESCRIPTION
**Pull request by hamz2a**
> [!WARNING]
> must be merged after https://github.com/OpenRailAssociation/osrd/pull/11819

I moved the files from the `core` module into a new crate called `editoast_core` and updated the imports in the other files accordingly to reflect this change. This refactor enhances modularity by separating `core` functionality into its own crate.

### CODE REVIEW

File: editoast/editoast_core/src/conflict_detection.rs [hunk 1]
```diff
@@ -43,7 +43,7 @@ pub struct WorkSchedulesRequest {
 }
 
 #[derive(Debug, Deserialize, ToSchema)]
-#[cfg_attr(test, derive(Serialize))]
+#[cfg_attr(feature = "testing", derive(Serialize))]
```
Comment (leovalais) [inline editoast/editoast_core/src/conflict_detection.rs]:
```suggestion
#[derive(Serialize)]
```

most definitions in core do not care about excessive `derive` (wrongfully though), but let's use consistency as a pretext to simplify things here...

File: editoast/editoast_core/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,40 @@
+[package]
+name = "editoast_core"
+version.workspace = true
+edition.workspace = true
+license.workspace = true
+
+[features]
+testing = []
```
Comment (leovalais) [inline editoast/editoast_core/Cargo.toml]:
```suggestion
mocking_client = []
```

**Review by leovalais — COMMENTED** [review]
Thanks! We're finally here!

How about naming simply naming the crate `core`? ping @woshilapin @flomonster @Tguisnet

We had a small discussion about this in a recent PR but I can't find it to link it here. Basically my point was that the `editoast_` prefix is cumbersome to write and therefore discourages the use of qualified paths in favor of prefixes in type names. And it doesn't really add much value either: we know we're in editoast.

File: editoast/editoast_core/src/conflict_detection.rs (hunk 1 above)
Comment (hamz2a) [reply editoast/editoast_core/src/conflict_detection.rs]: Done.

File: editoast/editoast_core/Cargo.toml (hunk 2 above)
Comment (hamz2a) [reply editoast/editoast_core/Cargo.toml]: Done.

**Review by leovalais — APPROVED** [review]
LGTM! Let's just settle on the name of the crate. Thanks!

**Comment by woshilapin** [comment]
Agreed on naming it only `core`. But let's not wait for renaming the other crates too (or we do everything in an independent PR).

File: editoast/Cargo.toml [hunk 3]
```diff
@@ -80,6 +87,9 @@ regex = "1.11"
 syn = "2.0"
 # 0.12.0 to 0.12.4 have weird timeout issues https://github.com/seanmonstar/reqwest/issues/2283
 # This bug was introduced between 0.12.0 and 0.12.3.
+approx = "0.5.1"
+async-std = { version = "1.13.0", features = ["attributes", "tokio1"] }
```
Comment (woshilapin) [inline editoast/Cargo.toml]: nitpick: please move these (`approx` and `async-std`) at the beginning of the section following alphabetical order. `taplo` usually do that by itself but the comment above make `taplo` fail 🤷 What you can do is remove the comment, run `taplo format` then put back the comment.

File: editoast/editoast_core/src/lib.rs [hunk 4]
```diff
@@ -115,7 +113,7 @@ impl CoreClient {
 ///
 /// For example:
 ///
-/// ```
+/// ```ignore
```
Comment (woshilapin) [inline editoast/editoast_core/src/lib.rs]: What is the reason for that? It might be interesting to add some comment about why this is ignored if there is a good reason.

File: editoast/editoast_core/src/pathfinding.rs [hunk 5]
```diff
@@ -183,7 +183,7 @@ impl From<editoast_schemas::infra::DirectionalTrackRange> for TrackRange {
     }
 }
 
-#[cfg(test)]
+#[cfg(feature = "mocking_client")]
```
Comment (woshilapin) [inline editoast/editoast_core/src/pathfinding.rs]: OK, interesting because this implementation doesn't relate to mocking client in any way. So changing from `#[cfg(test)]` to `#[cfg(feature = "mocking_client")]` doesn't make any sense.

This implementation is basically a convenience to write tests in a more concise way (see `views/path/projection.rs`). To me, the problem is that `TrackRange` is used both as an object of the API between `core` and `editoast`, and as an object of the API between `front` and `editoast`.

I see that you first named the feature `testing` which in this case would make more sense than `mocking_client`. I see 3 options:
- Should we have 2 features (cc @leovalais) `testing` and `mocking_client`?
- Another possibility would be to define 2 different `TrackRange` (probably the cleaner solution because there is semantically 2 objects)?
- A third possibility would be to replace this `impl` only and move it to where we need it (`views/path/projection.rs`), but it will probably need a free-function (because orphan-rule) and will be a pain to work with (because we use a special feature of `rstest` that works with `FromStr`).

**Comment by leovalais** [comment]
I'd prefer individual PRs though, much easier to review and to solve conflicts afterwards. Let's also wait for @flomonster and @Tguisnet opinion on the subject :)

File: editoast/editoast_core/src/pathfinding.rs [hunk 6]
```diff
@@ -220,7 +220,7 @@ impl std::str::FromStr for TrackRange {
 }
 
 impl TrackRange {
-    #[cfg(test)]
+    #[cfg(feature = "mocking_client")]
```
Comment (woshilapin) [inline editoast/editoast_core/src/pathfinding.rs]: Same for this, doesn't feel related to mocking in any way.

File: editoast/editoast_core/src/pathfinding.rs (hunk 5 above)
Comment (leovalais) [reply editoast/editoast_core/src/pathfinding.rs]: Good catch I missed this one.  I really don't like point 3.  I'd go with solution 2, but there already is 2 different track ranges: one in core, one in schemas.  Maybe we could move the parsing the schemas, gate it there, and rely on `into()` where necessary?

Comment (woshilapin) [reply editoast/editoast_core/src/pathfinding.rs]: I'm fine using the one in schemas for views (instead of using the one from core). We already have a bunch of schemas in views anyway.

File: editoast/Cargo.toml (hunk 3 above)
Comment (hamz2a) [reply editoast/Cargo.toml]: Done.

File: editoast/editoast_core/src/lib.rs (hunk 4 above)
Comment (hamz2a) [reply editoast/editoast_core/src/lib.rs]: I added a comment to explain the reason.

File: editoast/editoast_core/src/pathfinding.rs (hunk 5 above)
Comment (hamz2a) [reply editoast/editoast_core/src/pathfinding.rs]: If I got this right, we’ll move this implementation:
```rust
impl std::str::FromStr for TrackRange {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
```
to `editoast_schemas::infra::DirectionalTrackRange` (instead of `editoast_schemas::infra::TrackRange`, which doesn’t have direction). We’ll use it in the tests in `views`. Should we also add a `testing` feature for `editoast_schemas`?
If this is the plan, wouldn’t it be cleaner to do these changes in a separate PR and merge it before merging this PR, since they’re not related to creating `editoast_core`? @woshilapin @leovalais

Comment (leovalais) [reply editoast/editoast_core/src/pathfinding.rs]: My preference would be to to that in this pr to avoid merging a feature that doesn't make sense (thus making the history cleaner).  I'll let @woshilapin weigh on this 🥔

Comment (woshilapin) [reply editoast/editoast_core/src/pathfinding.rs]: I disagree with you on that @leovalais. I believe it does make sense to do that in a different PR. It's part of untangling `core` with the rest of the application: `views` is using objects from `core` it shouldn't (and splitting this `core` crate revealed that).

As to answer about the feature, a feature seems reasonable. I wonder if it should be named `testing` or something else? It basically provides a handy way to represent and write these directional track ranges, a bit like the crate `time` that provides `serde-human-readable`. So what about a feature named `human-readable` instead? I'm not absolutely attached to this name vs `testing` (I might have miss something in my reasoning) so I'll let you decide.

Comment (leovalais) [reply editoast/editoast_core/src/pathfinding.rs]: Okay I'm dumb... I re-read @hamz2a's message and clearly misread this part before:

> and merge it before merging this PR,

I thought you meant *after* this PR... So yeah sure it makes a lot of sense to do that before for teh reasons above. Sorry about the confusion 🤦

Comment (hamz2a) [reply editoast/editoast_core/src/pathfinding.rs]: https://github.com/OpenRailAssociation/osrd/pull/11819

File: editoast/editoast_core/src/pathfinding.rs (hunk 6 above)
Comment (hamz2a) [reply editoast/editoast_core/src/pathfinding.rs]: https://github.com/OpenRailAssociation/osrd/pull/11819

**Comment by hamz2a** [comment]
> How about naming simply naming the crate `core`? ping @woshilapin @flomonster @Tguisnet

@flomonster @Tguisnet

File: editoast/editoast_core/src/lib.rs (hunk 4 above)
Comment (woshilapin) [reply editoast/editoast_core/src/lib.rs]: Take a look in https://github.com/OpenRailAssociation/osrd/pull/11852.

File: editoast/editoast_core/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+serde_json.workspace = true
+thiserror.workspace = true
+tokio.workspace = true
+tracing.workspace = true
+tracing-opentelemetry.workspace = true
+url.workspace = true
+utoipa.workspace = true
+uuid.workspace = true
+
+[dev-dependencies]
```
Comment (woshilapin) [inline editoast/editoast_core/Cargo.toml]: You need to add `osrd_schemas` with the feature `testing` in this section. Take a look at https://github.com/OpenRailAssociation/osrd/pull/11852.

Comment (woshilapin) [inline editoast/editoast_core/Cargo.toml]: One of the problem of the `mocking_client` feature is that it is also used inside `editoast_core` itself to do some testing. If you try to run `cargo test` from inside `editoast_core`, you'll see that it fails because the feature is not there. Note that it works in the entire workspace because of how the `cargo` dependency resolver: it will find one version of the crate that works for the entire workspace. So if one crate depends on `editoast_core` without the feature, and another crate depends on `editoast_core` **with** the feature, the resolver will keep the version of `editoast_core` that works for both, so **with** the feature.
One way to make that work is to make `editoast_core` depends on itself in the `dev-dependencies`, with the feature activated:
```toml
editoast_core = { path = "./", features = ["mocking_client"] }
```

See https://github.com/OpenRailAssociation/osrd/pull/11852 which already provide the fix.

**Review by woshilapin — APPROVED** [review]
Some tiny improvements that I propose in https://github.com/OpenRailAssociation/osrd/pull/11852. But they are not mandatory for merging since they only bring:
- one more test, which is probably not that important
- some better dependency management, which are solving an edge case (running commands directly inside `editoast_core`)

File: editoast/editoast_core/Cargo.toml (hunk 7 above)
Comment (woshilapin) [reply editoast/editoast_core/Cargo.toml]: I merged the other PR #11852, so I'm resolving this thread. Please squash the commits from my PR in the right commits of your PR.

File: editoast/editoast_core/src/lib.rs (hunk 4 above)
Comment (woshilapin) [reply editoast/editoast_core/src/lib.rs]: I merged the other PR #11852, so I'm resolving this thread. Please squash the commits from my PR in the right commits of your PR.

File: editoast/editoast_core/Cargo.toml (hunk 7 above)
Comment (woshilapin) [reply editoast/editoast_core/Cargo.toml]: I merged the other PR #11852, so I'm resolving this thread. Please squash the commits from my PR in the right commits of your PR.

**Comment by hamz2a** [comment]
> How about naming simply naming the crate `core`? ping @woshilapin @flomonster @Tguisnet

Hi @flomonster, @Tguisnet 👋
Just wanted to ask again — can you please share what you think when you have time? Thank you!

**Comment by woshilapin** [comment]
ℹ️ [I discovered](https://github.com/rust-lang/rust/issues/140310) that the code in a `fn main` in a doctest might not be run. For example, look at [this code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=246516a10013d45fc448a8cf37869af0) which should panic but doesn't. Note that it really depends on the code around the `main` function: for example, [this works well](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=5ac510a3173fbd187c76c654acc0e37f).

So I pushed a slight modification in [here](https://github.com/OpenRailAssociation/osrd/compare/f9420ff0cbae5ff043bcbf89527613f9fe5f3af8..e47a881e284a).

**Comment by flomonster** [comment]
> How about naming simply naming the crate `core`? ping @woshilapin @flomonster @Tguisnet
> 
> We had a small discussion about this in a recent PR but I can't find it to link it here. Basically my point was that the `editoast_` prefix is cumbersome to write and therefore discourages the use of qualified paths in favor of prefixes in type names. And it doesn't really add much value either: we know we're in editoast.

I'm not opposed to it. I wonder if it's a convention, for example, axum and serde use a prefix.

**Comment by woshilapin** [comment]
> I'm not opposed to it. I wonder if it's a convention, for example, axum and serde use a prefix.

True, but they do publish them on crates.io, we do not.

Note that if the only problem with long names (`editoast_core`) instead of short name (`core`) is for having a more concise code, we can also [import our dependencies by renaming them](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#renaming-dependencies-in-cargotoml) like the following example in the `Cargo.toml`.

```toml
core = { package = "editoast_core", path = "./editoast_core/" }
```

Then we can directly use `core::` in our code instead of `editoast_core::`. It might be the best of both world (but a bit misleading if you don't know that it can be done?).

**Comment by leovalais** [comment]
> we can also [import our dependencies by renaming them](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#renaming-dependencies-in-cargotoml)

That solves my problem with noisy qualified paths, I didn't know we could do that! Since that was my main concern, I'm fine with both solutions. I still have a slight preference for renaming to have my folders alphabetized correctly, but maybe that's just the OCD speaking...

image

**Comment by hamz2a** [comment]
> we can also [import our dependencies by renaming them](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#renaming-dependencies-in-cargotoml) like the following example in the `Cargo.toml`

I also prefer changing the crate name directly instead of renaming it in `Cargo.toml`.

**Comment by woshilapin** [comment]
> I still have a slight preference for renaming to have my folders alphabetized correctly, but maybe that's just the OCD speaking...

Mmm, you also have a solution for that, the name of the folder doesn't have to be the name of the crate.

In `editoast/Cargo.toml`, you just do
```
[workspace]
members = ["core", "models", etc...]
```

Then in `editoast/core/Cargo.toml`, you do
```toml
[package]
name = "editoast_core"
```

🤷

**Comment by leovalais** [comment]
Ha! In this case I'm in favor of renaming the folder since we can always add the `editoast_` prefix later if we are to publish the crate at some point. One level of indirection less.

**Comment by Tguisnet** [comment]
I don't have a relevant opinion on this subject (no preference for me), so I don't mind the change for the naming, go for it.

**Comment by hamz2a** [comment]
As we discussed, I renamed `editoast_core` to `core_client`.

**Review by leovalais — APPROVED** [review]
Thanks for this work! 🚀
